group-coordinator/src/main/java/org/apache/kafka/coordinator/group/metrics/CoordinatorRuntimeMetrics.java

```diff
@@ -0,0 +1,100 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.coordinator.group.metrics;
+
+import org.apache.kafka.coordinator.group.runtime.CoordinatorRuntime.CoordinatorState;
+
+import java.util.function.Supplier;
+
+/**
+ * Used by the group and transaction coordinator runtimes, the metrics suite holds partition state gauges and sensors.
+ */
+public interface CoordinatorRuntimeMetrics extends AutoCloseable {
+    /**
+     * Returns the metrics group.
+     */
+    String metricsGroup();
+
+    /**
+     * Called when the partition state changes.
+     * @param oldState The old state.
+     * @param newState The new state to transition to.
+     */
+    void onPartitionStateChange(CoordinatorState oldState, CoordinatorState newState);
+
+    /**
+     * Record the partition load metric.
+     * @param startTimeMs The partition load start time.
+     * @param endTimeMs   The partition load end time.
+     */
+    void recordPartitionLoadSensor(long startTimeMs, long endTimeMs);
+
+    /**
+     * Get the counter for partitions in Loading state.
+     * Only used for testing.
+     */
+    long numPartitionsLoading();
+
+    /**
+     * Get the counter for partitions in Active state.
+     * Only used for testing.
+     */
+    long numPartitionsActive();
+
+    /**
+     * Get the counter for partitions in Failed state.
+     * Only used for testing.
+     */
+    long numPartitionsFailed();
+
+    /**
+     * Update the event queue time.
+     *
+     * @param durationMs The queue time.
+     */
+    void updateEventQueueTime(long durationMs);
+
+    /**
+     * Update the event queue processing time.
+     *
+     * @param durationMs The event processing time.
+     */
+    void updateEventQueueProcessingTime(long durationMs);
+
+    /**
+     * Record the failed event.
+     */
+    void recordFailedEvent();
+
+    /**
+     * Record the successful event.
+     */
+    void recordSuccessfulEvent();
+
+    /**
+     * Record the thread idle ratio.
+     * @param ratio The idle ratio.
+     */
+    void recordThreadIdleRatio(double ratio);
```

Comment:
Each EventProcessorThread records this ratio in `handleEvents()`. This will take into account time spent on context switching which is what we want, right?

Reply:
How is this reported then? Is there a metric per thread?

Reply:
We use a global sensor (`threadIdleRatioSensor`) that is shared across all threads. Each thread will record its own view of the idle ratio. So we will emit a single meter for all threads

Reply:
Created `MultiThreadedEventProcessorTest#testRecordThreadIdleRatioTwoThreads`. This doesn't test the exact idle ratio but confirms we record values with multiple threads.

Reply:
So we have an average an min for idle ratio. Is the minimum value just the lowest recorded at a given time frame. (And we don't have a way to identify which thread is the one with the lowest value)

Reply:
Right, we won't be able to differentiate the idle ratio for each individual thread. 

> So we have an average an min for idle ratio. Is the minimum value just the lowest recorded at a given time frame?

Yes, that's right